server/routes/user.js

```diff
@@ -1,31 +1,111 @@
 const express = require("express");
-const { check, validationResult } = require("express-validator");
+const { check, body, validationResult } = require("express-validator");
 const router = express.Router();
+const bcrypt = require("bcryptjs");
 
 const { User } = require("../database");
 
 router.post(
-  "/register",
+  "/signup",
   [
     check("name", "Name is required")
       .not()
       .isEmpty(),
-    check("email", "Please use a valid email address").isEmail()
+    check("email", "Please use a valid email address").isEmail(),
+    check("password", "Choose a password at least 6 characters long").isLength({
+      min: 6
+    }),
+    body("confirmPassword").custom((value, { req }) => {
+      if (value !== req.body.password) {
+        throw new Error("Password confirmation does not match password");
+      }
+      return true;
+    })
   ],
   async (req, res) => {
     const errors = validationResult(req);
-    if (errors.length) {
+    if (errors.length > 0) {
       return res.status(400).json({ errors: errors.array() });
     }
-    const { name, email } = req.body;
+
+    const { name, email, password } = req.body;
+
     try {
-      const user = new User({ name, email });
-      await user.save();
-      res.send(user);
+      //Check for an existing user
+      var user = await User.findOne({ email: email });
+      if (user) {
+        return res.status(400).json({ errors: ["Email already exists"] });
+      } else {
+        const newUser = new User({
+          name: name,
+          email: email,
+          password: password,
+          experience: {}
+        });
+
+        user = await newUser.save();
+
+        //On successful save, create payload and send response token with user
+        user.login(user._id, (err, token) => {
+          if (err) throw err;
+          res.status(201).json({
+            success: true,
+            token: "Bearer " + token,
+            user: user
+          });
+        });
+      }
     } catch (err) {
-      console.err(err.message);
+      console.log(err);
     }
   }
-);
+),
+  router.post(
+    "/login",
+    [
+      check("email", "Enter your account email address to login")
+        .not()
+        .isEmpty(),
+      check("password", "Please enter a password to login")
+        .not()
+        .isEmpty()
+    ],
+    async (req, res) => {
+      const errors = validationResult(req);
+
+      if (errors.length > 0) {
+        return res.status(400).json({ errors: errors.array() });
+      }
+
+      const { email, password } = req.body;
+
+      // Find if user exists
+      try {
+        var user = await User.findOne({ email });
+        if (!user) {
+          return res.status(404).json({ errors: ["Email not found"] });
```

Comment:
Looks good, but can we update the error message? I'm using it on the front end to show to the client, so "Email not found" will appear on the front end. Can we say "Cannot find a user with this email" or something a little more detailed then email not found.

Reply:
Definitely can.

On another note, we should standardize the way we send errors in JSON. Express-validator sends them as an array of objects each with a "msg" property. This might make it a mess to parse the errors on the frontend if we don't stick to one format so I'm going to update all error messages to be sent in a format matching those coming from express-validator. What do you guys think? @rkerins91 

### Example
```
errors: [
    {
      value: '1234',
      msg: 'Choose a password at least 6 characters long',
      param: 'password',
      location: 'body'
    },
    {
      value: 'asadf',
      msg: 'Password confirmation does not match password',
      param: 'confirmPassword',
      location: 'body'
    }
]
```

Reply:
As long as we have something standardized it should be fine.